### PR TITLE
Keep the filemtime for files when downloading them in a zip/tar

### LINKED ISSUE #2349: Files in the zip sent when downloading a folder should not have their date modified
### Steps to reproduce
1. Go to a shared link
2. Click on "Download"
3. Download the zip
1. Open the zip

### Expected behaviour

All files in the zip should have the same date as in the cloud

### Actual behaviour

All the dates are modified to the date the zip was created

### Server configuration
**Operating system**:

**Web server:**

**Database:**

**PHP version:** 7

**Nextcloud version:** (see Nextcloud admin page) 10.0.1

-------
@jancborchardt

### PR DESCRIPTION
**Pull request by nickvergessen**
Fix #2349 

@oparoz @LukasReschke 

Should we backport this? I don't think it's very critical?

### CODE REVIEW

**Comment by oparoz** [comment]
> Should we backport this? I don't think it's very critical?

Not critical, no, but very annoying to see all the dates reset. You lose all history and can't tell what's years old vs fresh.

**Comment by LukasReschke** [comment]
No backport from my PoV. Has the potential to break stuff. Never trust ZIP ;)

File: lib/private/Streamer.php [hunk 1]
```diff
@@ -93,11 +94,18 @@ public function addDirRecursive($dir, $internalDir='') {
 	 * @param int $size Filesize
 	 * @return bool $success
```
Comment (LukasReschke) [inline lib/private/Streamer.php]: PHPDoc is missing $time

File: lib/private/Streamer.php [hunk 2]
```diff
@@ -76,8 +76,9 @@ public function addDirRecursive($dir, $internalDir='') {
 			$file = $dir . '/' . $filename;
 			if(\OC\Files\Filesystem::is_file($file)) {
 				$filesize = \OC\Files\Filesystem::filesize($file);
+				$fileTime = \OC\Files\Filesystem::filemtime($file);
```
Comment (LukasReschke) [inline lib/private/Streamer.php]: Apparently also can return `false`. I'd advocate for catching such situations and then set it to the default `NULL`. (https://github.com/McNetic/PHPZipStreamer/blob/e57c198486242476587d04844084adbe8330581d/src/ZipStreamer.php#L202)

**Review by LukasReschke — CHANGES_REQUESTED** [review]

File: lib/private/Streamer.php (hunk 2 above)
Comment (nickvergessen) [inline lib/private/Streamer.php]: That's why we have `if ($time) {` in `addFileFromStream` 😉

**Comment by nickvergessen** [comment]
Fixed the doc, approve your review?

**Review by LukasReschke — APPROVED** [review]

**Comment by LukasReschke** [comment]
LGTM

**Comment by MorrisJobke** [comment]
Tested and works 👍

**Comment by oparoz** [comment]
Awesome guys, thanks :)

**Pull request merged**